drivers/clock_control/clock_control_r8a779f_cpg_mssr.c

```diff
@@ -13,72 +13,259 @@
+
+	switch (clk_info->module) {
+	case R8A779F_CLK_SD0CKCR1:
+		return 0;
+	case R8A779F_CLK_SD:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A779F_CLK_SD_STOP_BIT;
+		reg |= !enable * R8A779F_CLK_SD_STOP_BIT;
+		break;
+	case R8A779F_CLK_SDH:
+		reg = sys_read32(cfg->base_address + clk_info->offset);
+		reg &= ~R8A779F_CLK_SDH_STOP_BIT;
+		reg |= !enable * R8A779F_CLK_SDH_STOP_BIT;
+		break;
+	default:
+		ret = -ENOTSUP;
+		break;
+	}
+
+	if (!ret) {
+		rcar_cpg_write(cfg->base_address, clk_info->offset, reg);
+	}
+	return ret;
+}
+
+static int r8a779f_cpg_core_clock_endisable(const struct device *dev,
+					    struct rcar_cpg_clk *clk,
+					    bool enable)
+{
+	struct cpg_clk_info_table *clk_info;
+	const struct r8a779f_cpg_mssr_cfg *cfg = dev->config;
+	struct r8a779f_cpg_mssr_data *data = dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
+
+	clk_info = rcar_cpg_find_clk_info_by_module_id(dev, clk->domain, clk->module);
+	if (!clk_info) {
+		return -EINVAL;
+	}
+
+	if (enable) {
+		if ((int)clk->rate > 0) {
+			uintptr_t rate = clk->rate;
+
+			ret = rcar_cpg_set_rate(dev, (clock_control_subsys_t)clk,
+						(clock_control_subsys_rate_t)rate);
+			if (ret < 0) {
+				return ret;
+			}
+		}
+	}
+
+	key = k_spin_lock(&data->cmn.lock);
+	r8a779f_cpg_enable_disable_core(cfg, clk_info, enable);
+	k_spin_unlock(&data->cmn.lock, key);
+
+	return ret;
+}
+
 int r8a779f_cpg_mssr_start_stop(const struct device *dev,
 	clock_control_subsys_t sys, bool enable)
 {
-	const struct r8a779f_cpg_mssr_cfg *config = dev->config;
+	const struct r8a779f_cpg_mssr_cfg *config;
+	struct r8a779f_cpg_mssr_data *data;
 	struct rcar_cpg_clk *clk = (struct rcar_cpg_clk *)sys;
 	int ret = -EINVAL;
+	k_spinlock_key_t key;
+
+	if (!dev || !sys) {
+		return -EINVAL;
+	}
 
-	/* TODO: add support for CPG_CORE */
-	if (clk->domain != CPG_MOD) {
-		return ret;
+	if (clk->domain == CPG_CORE) {
+		return r8a779f_cpg_core_clock_endisable(dev, clk, enable);
 	}
-	return rcar_cpg_mstp_clock_endisable(config->base_address,
-			clk->module, enable);
+
+	config = dev->config;
+	data = dev->data;
+
+	key = k_spin_lock(&data->cmn.lock);
+	ret = rcar_cpg_mstp_clock_endisable(config->base_address, clk->module, enable);
+	k_spin_unlock(&data->cmn.lock, key);
+	return ret;
 }
 
 static int r8a779f_cpg_mssr_start(const struct device *dev,
-	clock_control_subsys_t sys)
+				  clock_control_subsys_t sys)
 {
 	return r8a779f_cpg_mssr_start_stop(dev, sys, true);
 }
 
 static int r8a779f_cpg_mssr_stop(const struct device *dev,
-	clock_control_subsys_t sys)
+				 clock_control_subsys_t sys)
 {
 	return r8a779f_cpg_mssr_start_stop(dev, sys, false);
 }
 
 static int r8a779f_cpg_mssr_init(const struct device *dev)
 {
-	ARG_UNUSED(dev);
+	rcar_cpg_build_clock_relationship(dev);
+	rcar_cpg_update_all_in_out_freq(dev);
 	return 0;
 }
 
+static uint32_t r8a779f_get_div_helper(uint32_t reg_val, uint32_t module)
+{
+	uint32_t divider = RCAR_CPG_NONE;
+
+	switch (module) {
```

Comment:
Lot of magics in this function. I think it should be described with comments or be a named `define`s